
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("tidyverse")
```

```{r}
install.packages("remotes")
```

```{r, message = FALSE}
#library("stats", quietly = T)
#library("SummarizedExperiment", quietly = T)
#library("PADOG", quietly = T)
library("knitr", quietly = T)
#library("tidyverse", quietly = T)
library("dplyr", quietly = T)
library("openxlsx", quietly = T)
library("tximport", quietly = T)
#library("ensembldb", quietly = T)
library("DESeq2", quietly = T)
#library("GenomicFeatures", quietly = T)
#library("RColorBrewer", quietly = T)
#library("gplots", quietly = T)
library("ggplot2", quietly = T)
library("tidyverse")
#library("ggrepel", quietly = T)
#library("tsne", quietly = T)
#library("umap", quietly = T)
#library("MatrixGenerics", quietly = T)
#library("BatchJobs", quietly = T)
#library("BiocParallel", quietly = T)
#library("biomaRt", quietly = T)
#library("pheatmap", quietly = T)
#library("randomcoloR", quietly = T)
#library("plotly", quietly = T)
#library("svglite", quietly = T)
#library("VennDiagram", quietly = T)
#library("pheatmap", quietly = T)
# library("ggsignif", quietly = T)
# library("gridExtra", quietly = T)
# library("cowplot", quietly = T)
# library("gtable", quietly = T)
# library("beeswarm", quietly = T)
# library("ggbeeswarm", quietly = T)
# library("scales", quietly = T)
# library("clusterProfiler", quietly = T)
# library("org.Hs.eg.db", quietly = T)
# library("AnnotationDbi", quietly = T)
# library("DOSE", quietly = T)
remotes::install_github("danoguevara/OmicsKit", ref = "new-functions", force = T)
library("OmicsKit")
```

## Set working directory

```{r}
require(knitr)
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "./")
sample_data <- "Samples_with_Counts.xlsx"
counts_dir <- "./Counts_Salmon-daniel"
```

# 2. Pre processing workflow

## Sample metadata

```{r}
sampledata <- data.frame(read.xlsx(sample_data))

# Manage numeric variables
# sampledata <- sampledata %>%
#   mutate(
#     bmi = ifelse(is.na(bmi), median(bmi, na.rm = TRUE), bmi),
#     HR = ifelse(is.na(HR), median(HR, na.rm = TRUE), HR),
#     weight = ifelse(is.na(weight), median(weight, na.rm = TRUE), weight),
#     waist = ifelse(is.na(waist), median(waist, na.rm = TRUE), waist),
#     hip_circ = ifelse(is.na(hip_circ), median(hip_circ, na.rm = TRUE), hip_circ)
#   )

# Convert factors
sampledata <- sampledata %>%
  mutate(
    # sex = factor(sex),
    # latent_TB = factor(latent_TB, levels = c(0, 1)),
    # age_cat = factor(age_cat, ordered = T, levels = c("20s", "30s", "40s", "50s", "60s")),
    group = factor(group, levels = c("Control", "TB"))
    # HR_cat = factor(HR_cat, levels = c("60s", "70s", "80s", "90s", "100s", "110s", "140s")),
    # weight_cat = factor(weight_cat, levels = c("40s", "50s", "60s", "70s", "80s", "90s")),
    # height_cat = factor(height_cat, levels = c(140, 150, 160, 170, 180))
  )

rownames(sampledata) <- sampledata$external_id

# Normal sample clustered with TB patients
#sampledata <- sampledata[-20, ] #301-1
```

## Managing Salmon count files

```{r}
count_files <- file.path(counts_dir, paste0(sampledata$external_id, "_quant.sf"))
names(count_files) <- sampledata$external_id

txi <- tximport(count_files, type="salmon", txOut=TRUE, countsFromAbundance="no")
counts.tx <- txi$counts
#counts.tx <- counts.tx[rowSums(counts.tx) > 0,]

sampledata$tx_counts <- colSums(counts.tx)

# Output
write.xlsx(sampledata, file = "samples_with_counts.xlsx", colNames = T, rowNames = F, append = F)
```

### Counts per sample

```{r}
my_colors <- c("#bbffff", "#ffb6c1")

p.counts <- ggplot(data = sampledata, aes(x = id, y = tx_counts, fill = group)) +
  theme_bw() + theme(panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank()) +
  geom_bar(stat="identity", color="black", linewidth = 0.3) + scale_fill_manual(values = my_colors) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(expand = expansion(mult = c(0, .01)), labels = scales::comma, breaks = seq(0,7e+7,by=0.5e+7)) +
  xlab("Sample ID") + ylab("Counts") + theme(axis.title=element_text(size=16))

p.counts

ggsave("plots/Counts-Transcript_per_Sample.jpg", p.counts, width = 10, height = 8, dpi = 300)
```
## Transcripts into genes

```{r}
tx2gene <- data.frame(read.xlsx("tx2gene_GRCh38v112.xlsx"))
txi.toGene <- summarizeToGene(txi, tx2gene[, c("transcriptID","geneID")])

# Add gene counts
counts.gene <- as.data.frame(txi.toGene$counts)
#counts.gene <- counts.gene[rowSums(counts.gene) > 0,]
sampledata$gene_counts <- colSums(counts.gene)

write.xlsx(sampledata, file = "samples_with_counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)

# Load geneID annotations
annotations <- c("symbol", "biotype", "chromosome", "gene_start", "gene_end", "gene_length", "description")
geneID.details <- tx2gene[,c("geneID", annotations)]
geneID.details <- geneID.details[!duplicated(geneID.details), ]

#######################
## Extra genes
### Remove NAs
for (i in 1:nrow(geneID.details)) {
  if (is.na(geneID.details$symbol[i]) | geneID.details$symbol[i] == "") {
    geneID.details$symbol[i] <- geneID.details$geneID[i]
  }
}
sum(is.na(geneID.details$symbol))
sum(geneID.details$symbol == "")
#######################

# Add annotations
counts.gene_annotations <- counts.gene
counts.gene_annotations$geneID <- rownames(counts.gene_annotations)
counts.gene_annotations <- OmicsKit::add_annotations(object = counts.gene_annotations,
                                                     reference = geneID.details,
                                                     variables = annotations)

write.xlsx(counts.gene_annotations, file = "Raw_Gene_Counts_Annotated.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)
```

## Filter out blacklists

```{r}
blacklist_nRibo <- data.frame(read.csv("blacklist_nuclear-ribo.tsv", sep = "\t", header = TRUE))
blacklist_mtRibo <- data.frame(read.csv("blacklist_mito-ribo.tsv", sep = "\t", header = TRUE))
blacklist_allRibo <- data.frame(read.csv("blacklist_all-ribo.tsv", sep = "\t", header = TRUE))

# Keep the genes in each blacklist
genecounts.nRibo <- counts.gene[rownames(counts.gene) %in% blacklist_nRibo[,1], ]
genecounts.mtRibo <- counts.gene[rownames(counts.gene) %in% blacklist_mtRibo[,1], ]
genecounts.filtered <- counts.gene[!(rownames(counts.gene) %in% blacklist_allRibo[,1]), ]

sampledata$nRibo <- colSums(genecounts.nRibo)
sampledata$mtRibo <- colSums(genecounts.mtRibo)
sampledata$filtered <- colSums(genecounts.filtered)

write.xlsx(sampledata, file = "samples_with_counts.xlsx", colNames = T, rowNames = F, append = F)

# Filter the txi object
keep <- !(rownames(txi.toGene$counts) %in% blacklist_allRibo[,1])

txi.toGene$abundance <- txi.toGene$abundance[keep, ]
txi.toGene$counts <- txi.toGene$counts[keep, ]
txi.toGene$length <- txi.toGene$length[keep, ]
```

### Counts per blacklist

```{r}
for (bl in c("nRibo", "mtRibo")) {
  p.sub <- ggplot(data = sampledata, aes(x = id, y = .data[[bl]], fill = group)) +
    theme_bw() + theme(panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank()) +
    geom_bar(stat="identity", color="black", linewidth = 0.3) + scale_fill_manual(values = my_colors) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    scale_y_continuous(expand = expansion(mult = c(0, .01)), labels = scales::comma) +
    xlab("Sample ID") + ylab(bl) + theme(axis.title=element_text(size=16))
  
  print(p.sub)
  
  ggsave(paste0("plots/Counts-Transcript_per_Sample_", bl, ".jpg"),
         p.sub, width = 10, height = 8, dpi = 300)
}

# Filtered
p.filt <- ggplot(data = sampledata, aes(x = id, y = filtered, fill = group)) +
  theme_bw() + theme(panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank()) +
  geom_bar(stat="identity", color="black", linewidth = 0.3) +
  scale_fill_manual(values = my_colors) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(expand = expansion(mult = c(0, .05)), labels = scales::comma, breaks = seq(0,7e+7,by=0.5e+7)) +
  xlab("Sample ID") + ylab("Filtered") + theme(axis.title=element_text(size=16))

p.filt

ggsave("plots/Counts-Transcript_per_Sample_filtered.jpg", p.filt,
       width = 10, height = 8, dpi = 300)
```

## Scaled TPMs

```{r}
# Calculate the TPM. The columns will now add 1 million.
gene.tpm <- tpm(counts.gene, counts.gene_annotations$gene_length)
gene.tpm <- data.frame(gene.tpm)

#write.xlsx(sampledata, file = "Samples_with_Counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)

gene.tpm.annotated <- OmicsKit::add_annotations(gene.tpm, geneID.details, variables = annotations)

write.xlsx(gene.tpm.annotated, file = "Gene_Counts_TPM.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)

# Keep the genes in each blacklist
geneTPM.nRibo <- gene.tpm[rownames(gene.tpm) %in% blacklist_nRibo[,1], ]
geneTPM.mtRibo <- gene.tpm[rownames(gene.tpm) %in% blacklist_mtRibo[,1], ]
geneTPM.allRibo <- gene.tpm[rownames(gene.tpm) %in% blacklist_allRibo[,1], ]

sampledata$TPM_nRibo <- colSums(geneTPM.nRibo[1:(length(geneTPM.nRibo))])
sampledata$TPM_mtRibo <- colSums(geneTPM.mtRibo[1:(length(geneTPM.nRibo))])
sampledata$TPM_allRibo <- colSums(geneTPM.allRibo[1:(length(geneTPM.nRibo))])

write.xlsx(sampledata, file = "samples_with_counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)
```

# 3. Differential Expression Analysis with DESeq2

```{r no filter}
# Note: Salmon counts are going to be rounded
ds.deseq <- DESeqDataSetFromTximport(txi.toGene, sampledata, ~group)
dea_formula <- paste(attr(terms(ds.deseq@design), "term.labels"), collapse = "_")

colData(ds.deseq)$group  <- factor(colData(ds.deseq)$group,
                                   levels = c("Control", "TB"))

# Note that read quantifications are rounded in DESeq, so Salmon counts < 0.5 are now 0
#nozero <- rowSums(counts(ds.deseq)) > 0
#ds.deseq <- ds.deseq[nozero, ] # Remove genes with zero counts
ds.deseq <- estimateSizeFactors(ds.deseq)

######################
# Normalized gene counts
normalized.counts <- as.data.frame(counts(ds.deseq, normalized = TRUE))

TB.ids <- sampledata$external_id[sampledata$group == "TB"]
C.ids <- sampledata$external_id[sampledata$group == "Control"]

normalized.counts$Mean.TB <- rowMeans(normalized.counts[, TB.ids])
normalized.counts$Mean.C <- rowMeans(normalized.counts[, C.ids])

# Adding gene name and description
normalized.counts.annotated <- OmicsKit::add_annotations(normalized.counts, geneID.details, variables = annotations)
write.xlsx(normalized.counts.annotated, file = "Normalized_Gene_counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)
######################

# Variance Stabilizing Transformation
transf.data <- varianceStabilizingTransformation(ds.deseq)

# Model matrix
ds.deseq <- DESeq(ds.deseq, modelMatrixType="standard", betaPrior=FALSE)

## DESeq2's dispersion estimates
#plotDispEsts(ds.deseq)

# This will show the possible comparisons, according to the design provided
resultsNames(ds.deseq)

cutoff_alpha <- 0.25 # Cut off p-value
cutoff_fold <- 1 # Cut off fold-change

res.TB_C <- results(ds.deseq, name = "group_TB_vs_Control",
                    altHypothesis="greaterAbs",
                    alpha = cutoff_alpha,
                    # independentFiltering = FALSE, # Automatic filtering
                    pAdjustMethod = "BH") # Benjamini Hochberg = FDR

summary(res.TB_C)
head(res.TB_C[(res.TB_C$padj < cutoff_alpha) & !is.na(res.TB_C$padj), ], n = 10)

add_annotations_mod <- function(object, reference, variables = NULL){

  # Gene IDs as rownames
  object$geneID <- rownames(object)
  
  # Match gene IDs and extract annotation variables
  index <- match(object$geneID, reference$geneID)
  object[, variables] <- reference[index, variables]

  return(object)
}

res.TB_C <- add_annotations_mod(res.TB_C, geneID.details, variables = annotations)
colnames(res.TB_C)[7] <- "ensembl"

write.xlsx(res.TB_C, file = paste0("results/res_TB_vs_C_", dea_formula, "_all.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Significant
res.TB_C.sig <- subset(res.TB_C, ((padj < cutoff_alpha) & !is.na(padj)))

write.xlsx(res.TB_C.sig, file = paste0("results/res_TB_vs_C_", dea_formula, "_padj<0.25.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Detectable differentially expressed genes
detect_list <- detect_filter(norm.counts = normalized.counts[, 1:19],
                             df.BvsA = res.TB_C.sig,
                             cutoffs = c(50, 50, 0),
                             samples.baseline = 11:19,
                             samples.condition1 = 1:10)
#detect_list$DetectGenes

res.TB_C.det <- res.TB_C.sig[rownames(res.TB_C.sig) %in% detect_list$DetectGenes, ]

write.xlsx(res.TB_C.det, file = paste0("results/res_TB_vs_C_", dea_formula, "padj<0.25_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.05, ], file = paste0("results/res_TB_vs_C_", dea_formula, "padj<0.05_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.01, ], file = paste0("results/res_TB_vs_C_", dea_formula, "padj<0.01_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

```
```{r}
# Number of genes before filtering
cat("Total genes before filtering:", nrow(counts(ds.deseq)), "\n")

# Number of genes after filtering zero-count genes
nozero <- rowSums(counts(ds.deseq)) > 0
cat("Total genes after filtering:", sum(nozero), "\n")

# Checking dimensions of input count matrix
cat("Number of genes (rows):", nrow(txi.toGene$counts), "\n")
cat("Number of samples (columns):", ncol(txi.toGene$counts), "\n")
```


```{r filter 0}
# Define the threshold for filtering
cat("Number of genes BEFORE filtering (original tximport object):", nrow(txi.toGene$counts), "\n")

filter_threshold <- 0
keep_genes <- rowMeans(txi.toGene$counts) > filter_threshold

cat("Number of genes with mean count < ", filter_threshold, " across all samples:", sum(!keep_genes), "\n")

txi.filtered <- txi.toGene
txi.filtered$counts    <- txi.toGene$counts[keep_genes, ]
txi.filtered$abundance <- txi.toGene$abundance[keep_genes, ]
txi.filtered$length    <- txi.toGene$length[keep_genes, ]

cat("Number of genes AFTER filtering (filtered tximport object):", nrow(txi.filtered$counts), "\n")

# Create DESeq2 dataset from the filtered counts
ds.deseq <- DESeqDataSetFromTximport(txi.filtered, sampledata, ~group)
dea_formula <- paste(attr(terms(ds.deseq@design), "term.labels"), collapse = "_")

colData(ds.deseq)$group  <- factor(colData(ds.deseq)$group,
                                   levels = c("Control", "TB"))

# Note that read quantifications are rounded in DESeq, so Salmon counts < 0.5 are now 0
#nozero <- rowSums(counts(ds.deseq)) > 0
#ds.deseq <- ds.deseq[nozero, ] # Remove genes with zero counts
ds.deseq <- estimateSizeFactors(ds.deseq)

######################
# Normalized gene counts
normalized.counts <- as.data.frame(counts(ds.deseq, normalized = TRUE))

TB.ids <- sampledata$external_id[sampledata$group == "TB"]
C.ids <- sampledata$external_id[sampledata$group == "Control"]

normalized.counts$Mean.TB <- rowMeans(normalized.counts[, TB.ids])
normalized.counts$Mean.C <- rowMeans(normalized.counts[, C.ids])

# Adding gene name and description
normalized.counts.annotated <- OmicsKit::add_annotations(normalized.counts, geneID.details, variables = annotations)
write.xlsx(normalized.counts.annotated, file = "Normalized_Gene_counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)
######################

# Variance Stabilizing Transformation
transf.data <- varianceStabilizingTransformation(ds.deseq)

# Model matrix
ds.deseq <- DESeq(ds.deseq, modelMatrixType="standard", betaPrior=FALSE)

## DESeq2's dispersion estimates
#plotDispEsts(ds.deseq)

# This will show the possible comparisons, according to the design provided
resultsNames(ds.deseq)

cutoff_alpha <- 0.25 # Cut off p-value
cutoff_fold <- 1 # Cut off fold-change

res.TB_C <- results(ds.deseq, name = "group_TB_vs_Control",
                    altHypothesis="greaterAbs",
                    alpha = cutoff_alpha,
                    # independentFiltering = FALSE, # Automatic filtering
                    pAdjustMethod = "BH") # Benjamini Hochberg = FDR

summary(res.TB_C)
head(res.TB_C[(res.TB_C$padj < cutoff_alpha) & !is.na(res.TB_C$padj), ], n = 10)

add_annotations_mod <- function(object, reference, variables = NULL){

  # Gene IDs as rownames
  object$geneID <- rownames(object)
  
  # Match gene IDs and extract annotation variables
  index <- match(object$geneID, reference$geneID)
  object[, variables] <- reference[index, variables]

  return(object)
}

res.TB_C <- add_annotations_mod(res.TB_C, geneID.details, variables = annotations)
colnames(res.TB_C)[7] <- "ensembl"

write.xlsx(res.TB_C, file = paste0("results-filter0/res_TB_vs_C_", dea_formula, "_all.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Significant
res.TB_C.sig <- subset(res.TB_C, ((padj < cutoff_alpha) & !is.na(padj)))

write.xlsx(res.TB_C.sig, file = paste0("results-filter0/res_TB_vs_C_", dea_formula, "_padj<0.25.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Detectable differentially expressed genes
detect_list <- detect_filter(norm.counts = normalized.counts[, 1:19],
                             df.BvsA = res.TB_C.sig,
                             cutoffs = c(50, 50, 0),
                             samples.baseline = 11:19,
                             samples.condition1 = 1:10)
#detect_list$DetectGenes

res.TB_C.det <- res.TB_C.sig[rownames(res.TB_C.sig) %in% detect_list$DetectGenes, ]

write.xlsx(res.TB_C.det, file = paste0("results-filter0/res_TB_vs_C_", dea_formula, "padj<0.25_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.05, ], file = paste0("results-filter0/res_TB_vs_C_", dea_formula, "padj<0.05_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.01, ], file = paste0("results-filter0/res_TB_vs_C_", dea_formula, "padj<0.01_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

```


```{r filter 1}
# Define the threshold for filtering
cat("Number of genes BEFORE filtering (original tximport object):", nrow(txi.toGene$counts), "\n")

filter_threshold <- 1
keep_genes <- rowMeans(txi.toGene$counts) >= filter_threshold

cat("Number of genes with mean count < ", filter_threshold, " across all samples:", sum(!keep_genes), "\n")

txi.filtered <- txi.toGene
txi.filtered$counts    <- txi.toGene$counts[keep_genes, ]
txi.filtered$abundance <- txi.toGene$abundance[keep_genes, ]
txi.filtered$length    <- txi.toGene$length[keep_genes, ]

cat("Number of genes AFTER filtering (filtered tximport object):", nrow(txi.filtered$counts), "\n")

# Create DESeq2 dataset from the filtered counts
ds.deseq <- DESeqDataSetFromTximport(txi.filtered, sampledata, ~group)
dea_formula <- paste(attr(terms(ds.deseq@design), "term.labels"), collapse = "_")

colData(ds.deseq)$group  <- factor(colData(ds.deseq)$group,
                                   levels = c("Control", "TB"))

# Note that read quantifications are rounded in DESeq, so Salmon counts < 0.5 are now 0
#nozero <- rowSums(counts(ds.deseq)) > 0
#ds.deseq <- ds.deseq[nozero, ] # Remove genes with zero counts
ds.deseq <- estimateSizeFactors(ds.deseq)

######################
# Normalized gene counts
normalized.counts <- as.data.frame(counts(ds.deseq, normalized = TRUE))

TB.ids <- sampledata$external_id[sampledata$group == "TB"]
C.ids <- sampledata$external_id[sampledata$group == "Control"]

normalized.counts$Mean.TB <- rowMeans(normalized.counts[, TB.ids])
normalized.counts$Mean.C <- rowMeans(normalized.counts[, C.ids])

# Adding gene name and description
normalized.counts.annotated <- OmicsKit::add_annotations(normalized.counts, geneID.details, variables = annotations)
write.xlsx(normalized.counts.annotated, file = "Normalized_Gene_counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)
######################

# Variance Stabilizing Transformation
transf.data <- varianceStabilizingTransformation(ds.deseq)

# Model matrix
ds.deseq <- DESeq(ds.deseq, modelMatrixType="standard", betaPrior=FALSE)

## DESeq2's dispersion estimates
#plotDispEsts(ds.deseq)

# This will show the possible comparisons, according to the design provided
resultsNames(ds.deseq)

cutoff_alpha <- 0.25 # Cut off p-value
cutoff_fold <- 1 # Cut off fold-change

res.TB_C <- results(ds.deseq, name = "group_TB_vs_Control",
                    altHypothesis="greaterAbs",
                    alpha = cutoff_alpha,
                    # independentFiltering = FALSE, # Automatic filtering
                    pAdjustMethod = "BH") # Benjamini Hochberg = FDR

summary(res.TB_C)
head(res.TB_C[(res.TB_C$padj < cutoff_alpha) & !is.na(res.TB_C$padj), ], n = 10)

add_annotations_mod <- function(object, reference, variables = NULL){

  # Gene IDs as rownames
  object$geneID <- rownames(object)
  
  # Match gene IDs and extract annotation variables
  index <- match(object$geneID, reference$geneID)
  object[, variables] <- reference[index, variables]

  return(object)
}

res.TB_C <- add_annotations_mod(res.TB_C, geneID.details, variables = annotations)
colnames(res.TB_C)[7] <- "ensembl"

write.xlsx(res.TB_C, file = paste0("results-filter1/res_TB_vs_C_", dea_formula, "_all.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Significant
res.TB_C.sig <- subset(res.TB_C, ((padj < cutoff_alpha) & !is.na(padj)))

write.xlsx(res.TB_C.sig, file = paste0("results-filter1/res_TB_vs_C_", dea_formula, "_padj<0.25.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Detectable differentially expressed genes
detect_list <- detect_filter(norm.counts = normalized.counts[, 1:19],
                             df.BvsA = res.TB_C.sig,
                             cutoffs = c(50, 50, 0),
                             samples.baseline = 11:19,
                             samples.condition1 = 1:10)
#detect_list$DetectGenes

res.TB_C.det <- res.TB_C.sig[rownames(res.TB_C.sig) %in% detect_list$DetectGenes, ]

write.xlsx(res.TB_C.det, file = paste0("results-filter1/res_TB_vs_C_", dea_formula, "padj<0.25_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.05, ], file = paste0("results-filter1/res_TB_vs_C_", dea_formula, "padj<0.05_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.01, ], file = paste0("results-filter1/res_TB_vs_C_", dea_formula, "padj<0.01_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

```


```{r filter 2}
# Define the threshold for filtering
cat("Number of genes BEFORE filtering (original tximport object):", nrow(txi.toGene$counts), "\n")

filter_threshold <- 2
keep_genes <- rowMeans(txi.toGene$counts) >= filter_threshold

cat("Number of genes with mean count < ", filter_threshold, " across all samples:", sum(!keep_genes), "\n")

txi.filtered <- txi.toGene
txi.filtered$counts    <- txi.toGene$counts[keep_genes, ]
txi.filtered$abundance <- txi.toGene$abundance[keep_genes, ]
txi.filtered$length    <- txi.toGene$length[keep_genes, ]

cat("Number of genes AFTER filtering (filtered tximport object):", nrow(txi.filtered$counts), "\n")

# Create DESeq2 dataset from the filtered counts
ds.deseq <- DESeqDataSetFromTximport(txi.filtered, sampledata, ~group)
dea_formula <- paste(attr(terms(ds.deseq@design), "term.labels"), collapse = "_")

colData(ds.deseq)$group  <- factor(colData(ds.deseq)$group,
                                   levels = c("Control", "TB"))

# Note that read quantifications are rounded in DESeq, so Salmon counts < 0.5 are now 0
#nozero <- rowSums(counts(ds.deseq)) > 0
#ds.deseq <- ds.deseq[nozero, ] # Remove genes with zero counts
ds.deseq <- estimateSizeFactors(ds.deseq)

######################
# Normalized gene counts
normalized.counts <- as.data.frame(counts(ds.deseq, normalized = TRUE))

TB.ids <- sampledata$external_id[sampledata$group == "TB"]
C.ids <- sampledata$external_id[sampledata$group == "Control"]

normalized.counts$Mean.TB <- rowMeans(normalized.counts[, TB.ids])
normalized.counts$Mean.C <- rowMeans(normalized.counts[, C.ids])

# Adding gene name and description
normalized.counts.annotated <- OmicsKit::add_annotations(normalized.counts, geneID.details, variables = annotations)
write.xlsx(normalized.counts.annotated, file = "Normalized_Gene_counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)
######################

# Variance Stabilizing Transformation
transf.data <- varianceStabilizingTransformation(ds.deseq)

# Model matrix
ds.deseq <- DESeq(ds.deseq, modelMatrixType="standard", betaPrior=FALSE)

## DESeq2's dispersion estimates
#plotDispEsts(ds.deseq)

# This will show the possible comparisons, according to the design provided
resultsNames(ds.deseq)

cutoff_alpha <- 0.25 # Cut off p-value
cutoff_fold <- 1 # Cut off fold-change

res.TB_C <- results(ds.deseq, name = "group_TB_vs_Control",
                    altHypothesis="greaterAbs",
                    alpha = cutoff_alpha,
                    # independentFiltering = FALSE, # Automatic filtering
                    pAdjustMethod = "BH") # Benjamini Hochberg = FDR

summary(res.TB_C)
head(res.TB_C[(res.TB_C$padj < cutoff_alpha) & !is.na(res.TB_C$padj), ], n = 10)

add_annotations_mod <- function(object, reference, variables = NULL){

  # Gene IDs as rownames
  object$geneID <- rownames(object)
  
  # Match gene IDs and extract annotation variables
  index <- match(object$geneID, reference$geneID)
  object[, variables] <- reference[index, variables]

  return(object)
}

res.TB_C <- add_annotations_mod(res.TB_C, geneID.details, variables = annotations)
colnames(res.TB_C)[7] <- "ensembl"

write.xlsx(res.TB_C, file = paste0("results-filter2/res_TB_vs_C_", dea_formula, "_all.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Significant
res.TB_C.sig <- subset(res.TB_C, ((padj < cutoff_alpha) & !is.na(padj)))

write.xlsx(res.TB_C.sig, file = paste0("results-filter2/res_TB_vs_C_", dea_formula, "_padj<0.25.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Detectable differentially expressed genes
detect_list <- detect_filter(norm.counts = normalized.counts[, 1:19],
                             df.BvsA = res.TB_C.sig,
                             cutoffs = c(50, 50, 0),
                             samples.baseline = 11:19,
                             samples.condition1 = 1:10)
#detect_list$DetectGenes

res.TB_C.det <- res.TB_C.sig[rownames(res.TB_C.sig) %in% detect_list$DetectGenes, ]

write.xlsx(res.TB_C.det, file = paste0("results-filter2/res_TB_vs_C_", dea_formula, "padj<0.25_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.05, ], file = paste0("results-filter2/res_TB_vs_C_", dea_formula, "padj<0.05_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.01, ], file = paste0("results-filter2/res_TB_vs_C_", dea_formula, "padj<0.01_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

```


```{r filter 3}
# Define the threshold for filtering
cat("Number of genes BEFORE filtering (original tximport object):", nrow(txi.toGene$counts), "\n")

filter_threshold <- 3
keep_genes <- rowMeans(txi.toGene$counts) >= filter_threshold

cat("Number of genes with mean count < ", filter_threshold, " across all samples:", sum(!keep_genes), "\n")

txi.filtered <- txi.toGene
txi.filtered$counts    <- txi.toGene$counts[keep_genes, ]
txi.filtered$abundance <- txi.toGene$abundance[keep_genes, ]
txi.filtered$length    <- txi.toGene$length[keep_genes, ]

cat("Number of genes AFTER filtering (filtered tximport object):", nrow(txi.filtered$counts), "\n")

# Create DESeq2 dataset from the filtered counts
ds.deseq <- DESeqDataSetFromTximport(txi.filtered, sampledata, ~group)
dea_formula <- paste(attr(terms(ds.deseq@design), "term.labels"), collapse = "_")

colData(ds.deseq)$group  <- factor(colData(ds.deseq)$group,
                                   levels = c("Control", "TB"))

# Note that read quantifications are rounded in DESeq, so Salmon counts < 0.5 are now 0
#nozero <- rowSums(counts(ds.deseq)) > 0
#ds.deseq <- ds.deseq[nozero, ] # Remove genes with zero counts
ds.deseq <- estimateSizeFactors(ds.deseq)

######################
# Normalized gene counts
normalized.counts <- as.data.frame(counts(ds.deseq, normalized = TRUE))

TB.ids <- sampledata$external_id[sampledata$group == "TB"]
C.ids <- sampledata$external_id[sampledata$group == "Control"]

normalized.counts$Mean.TB <- rowMeans(normalized.counts[, TB.ids])
normalized.counts$Mean.C <- rowMeans(normalized.counts[, C.ids])

# Adding gene name and description
normalized.counts.annotated <- OmicsKit::add_annotations(normalized.counts, geneID.details, variables = annotations)
write.xlsx(normalized.counts.annotated, file = "Normalized_Gene_counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)
######################

# Variance Stabilizing Transformation
transf.data <- varianceStabilizingTransformation(ds.deseq)

# Model matrix
ds.deseq <- DESeq(ds.deseq, modelMatrixType="standard", betaPrior=FALSE)

## DESeq2's dispersion estimates
#plotDispEsts(ds.deseq)

# This will show the possible comparisons, according to the design provided
resultsNames(ds.deseq)

cutoff_alpha <- 0.25 # Cut off p-value
cutoff_fold <- 1 # Cut off fold-change

res.TB_C <- results(ds.deseq, name = "group_TB_vs_Control",
                    altHypothesis="greaterAbs",
                    alpha = cutoff_alpha,
                    # independentFiltering = FALSE, # Automatic filtering
                    pAdjustMethod = "BH") # Benjamini Hochberg = FDR

summary(res.TB_C)
head(res.TB_C[(res.TB_C$padj < cutoff_alpha) & !is.na(res.TB_C$padj), ], n = 10)

add_annotations_mod <- function(object, reference, variables = NULL){

  # Gene IDs as rownames
  object$geneID <- rownames(object)
  
  # Match gene IDs and extract annotation variables
  index <- match(object$geneID, reference$geneID)
  object[, variables] <- reference[index, variables]

  return(object)
}

res.TB_C <- add_annotations_mod(res.TB_C, geneID.details, variables = annotations)
colnames(res.TB_C)[7] <- "ensembl"

write.xlsx(res.TB_C, file = paste0("results-filter3/res_TB_vs_C_", dea_formula, "_all.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Significant
res.TB_C.sig <- subset(res.TB_C, ((padj < cutoff_alpha) & !is.na(padj)))

write.xlsx(res.TB_C.sig, file = paste0("results-filter3/res_TB_vs_C_", dea_formula, "_padj<0.25.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Detectable differentially expressed genes
detect_list <- detect_filter(norm.counts = normalized.counts[, 1:19],
                             df.BvsA = res.TB_C.sig,
                             cutoffs = c(50, 50, 0),
                             samples.baseline = 11:19,
                             samples.condition1 = 1:10)
#detect_list$DetectGenes

res.TB_C.det <- res.TB_C.sig[rownames(res.TB_C.sig) %in% detect_list$DetectGenes, ]

write.xlsx(res.TB_C.det, file = paste0("results-filter3/res_TB_vs_C_", dea_formula, "padj<0.25_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.05, ], file = paste0("results-filter3/res_TB_vs_C_", dea_formula, "padj<0.05_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.01, ], file = paste0("results-filter3/res_TB_vs_C_", dea_formula, "padj<0.01_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

```



```{r filter 5}
# Define the threshold for filtering
cat("Number of genes BEFORE filtering (original tximport object):", nrow(txi.toGene$counts), "\n")

filter_threshold <- 5
keep_genes <- rowMeans(txi.toGene$counts) >= filter_threshold

cat("Number of genes with mean count < ", filter_threshold, " across all samples:", sum(!keep_genes), "\n")

txi.filtered <- txi.toGene
txi.filtered$counts    <- txi.toGene$counts[keep_genes, ]
txi.filtered$abundance <- txi.toGene$abundance[keep_genes, ]
txi.filtered$length    <- txi.toGene$length[keep_genes, ]

cat("Number of genes AFTER filtering (filtered tximport object):", nrow(txi.filtered$counts), "\n")

# Create DESeq2 dataset from the filtered counts
ds.deseq <- DESeqDataSetFromTximport(txi.filtered, sampledata, ~group)
dea_formula <- paste(attr(terms(ds.deseq@design), "term.labels"), collapse = "_")

colData(ds.deseq)$group  <- factor(colData(ds.deseq)$group,
                                   levels = c("Control", "TB"))

# Note that read quantifications are rounded in DESeq, so Salmon counts < 0.5 are now 0
#nozero <- rowSums(counts(ds.deseq)) > 0
#ds.deseq <- ds.deseq[nozero, ] # Remove genes with zero counts
ds.deseq <- estimateSizeFactors(ds.deseq)

######################
# Normalized gene counts
normalized.counts <- as.data.frame(counts(ds.deseq, normalized = TRUE))

TB.ids <- sampledata$external_id[sampledata$group == "TB"]
C.ids <- sampledata$external_id[sampledata$group == "Control"]

normalized.counts$Mean.TB <- rowMeans(normalized.counts[, TB.ids])
normalized.counts$Mean.C <- rowMeans(normalized.counts[, C.ids])

# Adding gene name and description
normalized.counts.annotated <- OmicsKit::add_annotations(normalized.counts, geneID.details, variables = annotations)
write.xlsx(normalized.counts.annotated, file = "Normalized_Gene_counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)
######################

# Variance Stabilizing Transformation
transf.data <- varianceStabilizingTransformation(ds.deseq)

# Model matrix
ds.deseq <- DESeq(ds.deseq, modelMatrixType="standard", betaPrior=FALSE)

## DESeq2's dispersion estimates
#plotDispEsts(ds.deseq)

# This will show the possible comparisons, according to the design provided
resultsNames(ds.deseq)

cutoff_alpha <- 0.25 # Cut off p-value
cutoff_fold <- 1 # Cut off fold-change

res.TB_C <- results(ds.deseq, name = "group_TB_vs_Control",
                    altHypothesis="greaterAbs",
                    alpha = cutoff_alpha,
                    # independentFiltering = FALSE, # Automatic filtering
                    pAdjustMethod = "BH") # Benjamini Hochberg = FDR

summary(res.TB_C)
head(res.TB_C[(res.TB_C$padj < cutoff_alpha) & !is.na(res.TB_C$padj), ], n = 10)

add_annotations_mod <- function(object, reference, variables = NULL){

  # Gene IDs as rownames
  object$geneID <- rownames(object)
  
  # Match gene IDs and extract annotation variables
  index <- match(object$geneID, reference$geneID)
  object[, variables] <- reference[index, variables]

  return(object)
}

res.TB_C <- add_annotations_mod(res.TB_C, geneID.details, variables = annotations)
colnames(res.TB_C)[7] <- "ensembl"

write.xlsx(res.TB_C, file = paste0("results-filter5/res_TB_vs_C_", dea_formula, "_all.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Significant
res.TB_C.sig <- subset(res.TB_C, ((padj < cutoff_alpha) & !is.na(padj)))

write.xlsx(res.TB_C.sig, file = paste0("results-filter5/res_TB_vs_C_", dea_formula, "_padj<0.25.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Detectable differentially expressed genes
detect_list <- detect_filter(norm.counts = normalized.counts[, 1:19],
                             df.BvsA = res.TB_C.sig,
                             cutoffs = c(50, 50, 0),
                             samples.baseline = 11:19,
                             samples.condition1 = 1:10)
#detect_list$DetectGenes

res.TB_C.det <- res.TB_C.sig[rownames(res.TB_C.sig) %in% detect_list$DetectGenes, ]

write.xlsx(res.TB_C.det, file = paste0("results-filter5/res_TB_vs_C_", dea_formula, "padj<0.25_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.05, ], file = paste0("results-filter5/res_TB_vs_C_", dea_formula, "padj<0.05_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.01, ], file = paste0("results-filter5/res_TB_vs_C_", dea_formula, "padj<0.01_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
```


```{r filter 10}
# Define the threshold for filtering
cat("Number of genes BEFORE filtering (original tximport object):", nrow(txi.toGene$counts), "\n")

filter_threshold <- 10
keep_genes <- rowMeans(txi.toGene$counts) >= filter_threshold

cat("Number of genes with mean count < ", filter_threshold, " across all samples:", sum(!keep_genes), "\n")

txi.filtered <- txi.toGene
txi.filtered$counts    <- txi.toGene$counts[keep_genes, ]
txi.filtered$abundance <- txi.toGene$abundance[keep_genes, ]
txi.filtered$length    <- txi.toGene$length[keep_genes, ]

cat("Number of genes AFTER filtering (filtered tximport object):", nrow(txi.filtered$counts), "\n")

# Create DESeq2 dataset from the filtered counts
ds.deseq <- DESeqDataSetFromTximport(txi.filtered, sampledata, ~group)
dea_formula <- paste(attr(terms(ds.deseq@design), "term.labels"), collapse = "_")

colData(ds.deseq)$group  <- factor(colData(ds.deseq)$group,
                                   levels = c("Control", "TB"))

# Note that read quantifications are rounded in DESeq, so Salmon counts < 0.5 are now 0
#nozero <- rowSums(counts(ds.deseq)) > 0
#ds.deseq <- ds.deseq[nozero, ] # Remove genes with zero counts
ds.deseq <- estimateSizeFactors(ds.deseq)

######################
# Normalized gene counts
normalized.counts <- as.data.frame(counts(ds.deseq, normalized = TRUE))

TB.ids <- sampledata$external_id[sampledata$group == "TB"]
C.ids <- sampledata$external_id[sampledata$group == "Control"]

normalized.counts$Mean.TB <- rowMeans(normalized.counts[, TB.ids])
normalized.counts$Mean.C <- rowMeans(normalized.counts[, C.ids])

# Adding gene name and description
normalized.counts.annotated <- OmicsKit::add_annotations(normalized.counts, geneID.details, variables = annotations)
write.xlsx(normalized.counts.annotated, file = "Normalized_Gene_counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)
######################

# Variance Stabilizing Transformation
transf.data <- varianceStabilizingTransformation(ds.deseq)

# Model matrix
ds.deseq <- DESeq(ds.deseq, modelMatrixType="standard", betaPrior=FALSE)

## DESeq2's dispersion estimates
#plotDispEsts(ds.deseq)

# This will show the possible comparisons, according to the design provided
resultsNames(ds.deseq)

cutoff_alpha <- 0.25 # Cut off p-value
cutoff_fold <- 1 # Cut off fold-change

res.TB_C <- results(ds.deseq, name = "group_TB_vs_Control",
                    altHypothesis="greaterAbs",
                    alpha = cutoff_alpha,
                    # independentFiltering = FALSE, # Automatic filtering
                    pAdjustMethod = "BH") # Benjamini Hochberg = FDR

summary(res.TB_C)
head(res.TB_C[(res.TB_C$padj < cutoff_alpha) & !is.na(res.TB_C$padj), ], n = 10)

add_annotations_mod <- function(object, reference, variables = NULL){

  # Gene IDs as rownames
  object$geneID <- rownames(object)
  
  # Match gene IDs and extract annotation variables
  index <- match(object$geneID, reference$geneID)
  object[, variables] <- reference[index, variables]

  return(object)
}

res.TB_C <- add_annotations_mod(res.TB_C, geneID.details, variables = annotations)
colnames(res.TB_C)[7] <- "ensembl"

write.xlsx(res.TB_C, file = paste0("results-filter10/res_TB_vs_C_", dea_formula, "_all.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Significant
res.TB_C.sig <- subset(res.TB_C, ((padj < cutoff_alpha) & !is.na(padj)))

write.xlsx(res.TB_C.sig, file = paste0("results-filter10/res_TB_vs_C_", dea_formula, "_padj<0.25.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Detectable differentially expressed genes
detect_list <- detect_filter(norm.counts = normalized.counts[, 1:19],
                             df.BvsA = res.TB_C.sig,
                             cutoffs = c(50, 50, 0),
                             samples.baseline = 11:19,
                             samples.condition1 = 1:10)
#detect_list$DetectGenes

res.TB_C.det <- res.TB_C.sig[rownames(res.TB_C.sig) %in% detect_list$DetectGenes, ]

write.xlsx(res.TB_C.det, file = paste0("results-filter10/res_TB_vs_C_", dea_formula, "padj<0.25_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.05, ], file = paste0("results-filter10/res_TB_vs_C_", dea_formula, "padj<0.05_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.01, ], file = paste0("results-filter10/res_TB_vs_C_", dea_formula, "padj<0.01_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
```


```{r filter20}
# Define the threshold for filtering
cat("Number of genes BEFORE filtering (original tximport object):", nrow(txi.toGene$counts), "\n")

filter_threshold <- 20
keep_genes <- rowMeans(txi.toGene$counts) >= filter_threshold

cat("Number of genes with mean count < ", filter_threshold, " across all samples:", sum(!keep_genes), "\n")

txi.filtered <- txi.toGene
txi.filtered$counts    <- txi.toGene$counts[keep_genes, ]
txi.filtered$abundance <- txi.toGene$abundance[keep_genes, ]
txi.filtered$length    <- txi.toGene$length[keep_genes, ]

cat("Number of genes AFTER filtering (filtered tximport object):", nrow(txi.filtered$counts), "\n")

# Create DESeq2 dataset from the filtered counts
ds.deseq <- DESeqDataSetFromTximport(txi.filtered, sampledata, ~group)
dea_formula <- paste(attr(terms(ds.deseq@design), "term.labels"), collapse = "_")

colData(ds.deseq)$group  <- factor(colData(ds.deseq)$group,
                                   levels = c("Control", "TB"))

# Note that read quantifications are rounded in DESeq, so Salmon counts < 0.5 are now 0
#nozero <- rowSums(counts(ds.deseq)) > 0
#ds.deseq <- ds.deseq[nozero, ] # Remove genes with zero counts
ds.deseq <- estimateSizeFactors(ds.deseq)

######################
# Normalized gene counts
normalized.counts <- as.data.frame(counts(ds.deseq, normalized = TRUE))

TB.ids <- sampledata$external_id[sampledata$group == "TB"]
C.ids <- sampledata$external_id[sampledata$group == "Control"]

normalized.counts$Mean.TB <- rowMeans(normalized.counts[, TB.ids])
normalized.counts$Mean.C <- rowMeans(normalized.counts[, C.ids])

# Adding gene name and description
normalized.counts.annotated <- OmicsKit::add_annotations(normalized.counts, geneID.details, variables = annotations)
write.xlsx(normalized.counts.annotated, file = "Normalized_Gene_counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)
######################

# Variance Stabilizing Transformation
transf.data <- varianceStabilizingTransformation(ds.deseq)

# Model matrix
ds.deseq <- DESeq(ds.deseq, modelMatrixType="standard", betaPrior=FALSE)

## DESeq2's dispersion estimates
#plotDispEsts(ds.deseq)

# This will show the possible comparisons, according to the design provided
resultsNames(ds.deseq)

cutoff_alpha <- 0.25 # Cut off p-value
cutoff_fold <- 1 # Cut off fold-change

res.TB_C <- results(ds.deseq, name = "group_TB_vs_Control",
                    altHypothesis="greaterAbs",
                    alpha = cutoff_alpha,
                    # independentFiltering = FALSE, # Automatic filtering
                    pAdjustMethod = "BH") # Benjamini Hochberg = FDR

summary(res.TB_C)
head(res.TB_C[(res.TB_C$padj < cutoff_alpha) & !is.na(res.TB_C$padj), ], n = 10)

add_annotations_mod <- function(object, reference, variables = NULL){

  # Gene IDs as rownames
  object$geneID <- rownames(object)
  
  # Match gene IDs and extract annotation variables
  index <- match(object$geneID, reference$geneID)
  object[, variables] <- reference[index, variables]

  return(object)
}

res.TB_C <- add_annotations_mod(res.TB_C, geneID.details, variables = annotations)
colnames(res.TB_C)[7] <- "ensembl"

write.xlsx(res.TB_C, file = paste0("results-filter20/res_TB_vs_C_", dea_formula, "_all.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Significant
res.TB_C.sig <- subset(res.TB_C, ((padj < cutoff_alpha) & !is.na(padj)))

write.xlsx(res.TB_C.sig, file = paste0("results-filter20/res_TB_vs_C_", dea_formula, "_padj<0.25.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Detectable differentially expressed genes
detect_list <- detect_filter(norm.counts = normalized.counts[, 1:19],
                             df.BvsA = res.TB_C.sig,
                             cutoffs = c(50, 50, 0),
                             samples.baseline = 11:19,
                             samples.condition1 = 1:10)
#detect_list$DetectGenes

res.TB_C.det <- res.TB_C.sig[rownames(res.TB_C.sig) %in% detect_list$DetectGenes, ]

write.xlsx(res.TB_C.det, file = paste0("results-filter20/res_TB_vs_C_", dea_formula, "padj<0.25_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.05, ], file = paste0("results-filter20/res_TB_vs_C_", dea_formula, "padj<0.05_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.01, ], file = paste0("results-filter20/res_TB_vs_C_", dea_formula, "padj<0.01_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
```

```{r filter50}
# Define the threshold for filtering
cat("Number of genes BEFORE filtering (original tximport object):", nrow(txi.toGene$counts), "\n")

filter_threshold <- 50
keep_genes <- rowMeans(txi.toGene$counts) >= filter_threshold

cat("Number of genes with mean count < ", filter_threshold, " across all samples:", sum(!keep_genes), "\n")

txi.filtered <- txi.toGene
txi.filtered$counts    <- txi.toGene$counts[keep_genes, ]
txi.filtered$abundance <- txi.toGene$abundance[keep_genes, ]
txi.filtered$length    <- txi.toGene$length[keep_genes, ]

cat("Number of genes AFTER filtering (filtered tximport object):", nrow(txi.filtered$counts), "\n")

# Create DESeq2 dataset from the filtered counts
ds.deseq <- DESeqDataSetFromTximport(txi.filtered, sampledata, ~group)
dea_formula <- paste(attr(terms(ds.deseq@design), "term.labels"), collapse = "_")

colData(ds.deseq)$group  <- factor(colData(ds.deseq)$group,
                                   levels = c("Control", "TB"))

# Note that read quantifications are rounded in DESeq, so Salmon counts < 0.5 are now 0
#nozero <- rowSums(counts(ds.deseq)) > 0
#ds.deseq <- ds.deseq[nozero, ] # Remove genes with zero counts
ds.deseq <- estimateSizeFactors(ds.deseq)

######################
# Normalized gene counts
normalized.counts <- as.data.frame(counts(ds.deseq, normalized = TRUE))

TB.ids <- sampledata$external_id[sampledata$group == "TB"]
C.ids <- sampledata$external_id[sampledata$group == "Control"]

normalized.counts$Mean.TB <- rowMeans(normalized.counts[, TB.ids])
normalized.counts$Mean.C <- rowMeans(normalized.counts[, C.ids])

# Adding gene name and description
normalized.counts.annotated <- OmicsKit::add_annotations(normalized.counts, geneID.details, variables = annotations)
write.xlsx(normalized.counts.annotated, file = "Normalized_Gene_counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)
######################

# Variance Stabilizing Transformation
transf.data <- varianceStabilizingTransformation(ds.deseq)

# Model matrix
ds.deseq <- DESeq(ds.deseq, modelMatrixType="standard", betaPrior=FALSE)

## DESeq2's dispersion estimates
#plotDispEsts(ds.deseq)

# This will show the possible comparisons, according to the design provided
resultsNames(ds.deseq)

cutoff_alpha <- 0.25 # Cut off p-value
cutoff_fold <- 1 # Cut off fold-change

res.TB_C <- results(ds.deseq, name = "group_TB_vs_Control",
                    altHypothesis="greaterAbs",
                    alpha = cutoff_alpha,
                    # independentFiltering = FALSE, # Automatic filtering
                    pAdjustMethod = "BH") # Benjamini Hochberg = FDR

summary(res.TB_C)
head(res.TB_C[(res.TB_C$padj < cutoff_alpha) & !is.na(res.TB_C$padj), ], n = 10)

add_annotations_mod <- function(object, reference, variables = NULL){

  # Gene IDs as rownames
  object$geneID <- rownames(object)
  
  # Match gene IDs and extract annotation variables
  index <- match(object$geneID, reference$geneID)
  object[, variables] <- reference[index, variables]

  return(object)
}

res.TB_C <- add_annotations_mod(res.TB_C, geneID.details, variables = annotations)
colnames(res.TB_C)[7] <- "ensembl"

write.xlsx(res.TB_C, file = paste0("results-filter50/res_TB_vs_C_", dea_formula, "_all.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Significant
res.TB_C.sig <- subset(res.TB_C, ((padj < cutoff_alpha) & !is.na(padj)))

write.xlsx(res.TB_C.sig, file = paste0("results-filter50/res_TB_vs_C_", dea_formula, "_padj<0.25.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Detectable differentially expressed genes
detect_list <- detect_filter(norm.counts = normalized.counts[, 1:19],
                             df.BvsA = res.TB_C.sig,
                             cutoffs = c(50, 50, 0),
                             samples.baseline = 11:19,
                             samples.condition1 = 1:10)
#detect_list$DetectGenes

res.TB_C.det <- res.TB_C.sig[rownames(res.TB_C.sig) %in% detect_list$DetectGenes, ]

write.xlsx(res.TB_C.det, file = paste0("results-filter50/res_TB_vs_C_", dea_formula, "padj<0.25_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.05, ], file = paste0("results-filter50/res_TB_vs_C_", dea_formula, "padj<0.05_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.01, ], file = paste0("results-filter50/res_TB_vs_C_", dea_formula, "padj<0.01_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
```

```{r filter100}
# Define the threshold for filtering
cat("Number of genes BEFORE filtering (original tximport object):", nrow(txi.toGene$counts), "\n")

filter_threshold <- 100
keep_genes <- rowMeans(txi.toGene$counts) >= filter_threshold

cat("Number of genes with mean count < ", filter_threshold, " across all samples:", sum(!keep_genes), "\n")

txi.filtered <- txi.toGene
txi.filtered$counts    <- txi.toGene$counts[keep_genes, ]
txi.filtered$abundance <- txi.toGene$abundance[keep_genes, ]
txi.filtered$length    <- txi.toGene$length[keep_genes, ]

cat("Number of genes AFTER filtering (filtered tximport object):", nrow(txi.filtered$counts), "\n")

# Create DESeq2 dataset from the filtered counts
ds.deseq <- DESeqDataSetFromTximport(txi.filtered, sampledata, ~group)
dea_formula <- paste(attr(terms(ds.deseq@design), "term.labels"), collapse = "_")

colData(ds.deseq)$group  <- factor(colData(ds.deseq)$group,
                                   levels = c("Control", "TB"))

# Note that read quantifications are rounded in DESeq, so Salmon counts < 0.5 are now 0
#nozero <- rowSums(counts(ds.deseq)) > 0
#ds.deseq <- ds.deseq[nozero, ] # Remove genes with zero counts
ds.deseq <- estimateSizeFactors(ds.deseq)

######################
# Normalized gene counts
normalized.counts <- as.data.frame(counts(ds.deseq, normalized = TRUE))

TB.ids <- sampledata$external_id[sampledata$group == "TB"]
C.ids <- sampledata$external_id[sampledata$group == "Control"]

normalized.counts$Mean.TB <- rowMeans(normalized.counts[, TB.ids])
normalized.counts$Mean.C <- rowMeans(normalized.counts[, C.ids])

# Adding gene name and description
normalized.counts.annotated <- OmicsKit::add_annotations(normalized.counts, geneID.details, variables = annotations)
write.xlsx(normalized.counts.annotated, file = "Normalized_Gene_counts.xlsx", colNames = T, rowNames = F, append = F, overwrite = T)
######################

# Variance Stabilizing Transformation
transf.data <- varianceStabilizingTransformation(ds.deseq)

# Model matrix
ds.deseq <- DESeq(ds.deseq, modelMatrixType="standard", betaPrior=FALSE)

## DESeq2's dispersion estimates
#plotDispEsts(ds.deseq)

# This will show the possible comparisons, according to the design provided
resultsNames(ds.deseq)

cutoff_alpha <- 0.25 # Cut off p-value
cutoff_fold <- 1 # Cut off fold-change

res.TB_C <- results(ds.deseq, name = "group_TB_vs_Control",
                    altHypothesis="greaterAbs",
                    alpha = cutoff_alpha,
                    # independentFiltering = FALSE, # Automatic filtering
                    pAdjustMethod = "BH") # Benjamini Hochberg = FDR

summary(res.TB_C)
head(res.TB_C[(res.TB_C$padj < cutoff_alpha) & !is.na(res.TB_C$padj), ], n = 10)

add_annotations_mod <- function(object, reference, variables = NULL){

  # Gene IDs as rownames
  object$geneID <- rownames(object)
  
  # Match gene IDs and extract annotation variables
  index <- match(object$geneID, reference$geneID)
  object[, variables] <- reference[index, variables]

  return(object)
}

res.TB_C <- add_annotations_mod(res.TB_C, geneID.details, variables = annotations)
colnames(res.TB_C)[7] <- "ensembl"

write.xlsx(res.TB_C, file = paste0("results-filter100/res_TB_vs_C_", dea_formula, "_all.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Significant
res.TB_C.sig <- subset(res.TB_C, ((padj < cutoff_alpha) & !is.na(padj)))

write.xlsx(res.TB_C.sig, file = paste0("results-filter100/res_TB_vs_C_", dea_formula, "_padj<0.25.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)

# Detectable differentially expressed genes
detect_list <- detect_filter(norm.counts = normalized.counts[, 1:19],
                             df.BvsA = res.TB_C.sig,
                             cutoffs = c(50, 50, 0),
                             samples.baseline = 11:19,
                             samples.condition1 = 1:10)
#detect_list$DetectGenes

res.TB_C.det <- res.TB_C.sig[rownames(res.TB_C.sig) %in% detect_list$DetectGenes, ]

write.xlsx(res.TB_C.det, file = paste0("results-filter100/res_TB_vs_C_", dea_formula, "padj<0.25_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.05, ], file = paste0("results-filter100/res_TB_vs_C_", dea_formula, "padj<0.05_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
write.xlsx(res.TB_C.det[res.TB_C.det$padj < 0.01, ], file = paste0("results-filter100/res_TB_vs_C_", dea_formula, "padj<0.01_detectable.xlsx"),
           colNames = TRUE, rowNames = FALSE, append = FALSE, overwrite = TRUE)
```


```{r}
binomial <- read.xlsx("./binomal_comparisons.xlsx")
binomial
```

```{r data wranlgin}
binomial_tidy <- binomial %>% 
  mutate(situation = fct_inorder(situation))

```


```{r plot}

plot_binomial <- ggplot(binomial_tidy, aes(x=situation, y=`DEGs.from.deseq2.(p<0.05)`, group=1)) +
  geom_jitter(width=0, size=2,height = 0) +
  geom_line()+
  theme(
  axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
  axis.text.y = element_text(size = 12),
  axis.title.y = element_text(size = 14),
  axis.title.x = element_text(size = 14),
  legend.position = "top",
  plot.title.position = "plot",
  plot.title = element_text(hjust = 0.5),
  panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),
  panel.background = element_blank(),
  axis.line = element_line(colour = "black", size = 0.5),
  legend.key = element_rect(fill = "white", colour = NA),
  strip.text = element_text(face = "bold", size = rel(1))
)
plot_binomial

ggsave("./binomial_plot.png")
```




